lib/logflare/backends/ingest_event_queue.ex

```diff
@@ -0,0 +1,288 @@
+  """
+  def take_pending({%Source{} = source, nil}, n), do: take_pending({source.id, nil}, n)
+  def take_pending({%Source{id: sid}, %Backend{id: bid}}, n), do: take_pending({sid, bid}, n)
+
+  def take_pending({sid, _bid} = sid_bid, 0) when is_integer(sid), do: {:ok, []}
+
+  def take_pending({sid, _bid} = sid_bid, n) when is_integer(sid) and is_integer(n) do
+    ms =
+      Ex2ms.fun do
+        {_event_id, :pending, event} -> event
+      end
+
+    with tid when tid != nil <- get_tid(sid_bid),
+         size when is_integer(size) <- :ets.info(tid, :size),
+         {taken, _cont} <- :ets.select(tid, ms, min(n, max(size, 1))) do
+      {:ok, taken}
+    else
+      nil -> {:error, :not_initialized}
+      :"$end_of_table" -> {:ok, []}
+    end
+  end
+
+  @doc """
+  Truncates a given table
+  """
+  def truncate({%Source{} = source, nil}, filter, n), do: truncate({source.id, nil}, filter, n)
+
+  def truncate({%Source{id: sid}, %Backend{id: bid}}, filter, n),
+    do: truncate({sid, bid}, filter, n)
+
+  def truncate({sid, _bid} = sid_bid, :all, 0) when is_integer(sid) do
+    # drop all objects
+    with tid when tid != nil <- get_tid(sid_bid) do
+      :ets.delete_all_objects(tid)
+      :ok
+    else
+      nil -> {:error, :not_initialized}
+    end
+  end
+
+  def truncate({sid, _bid} = sid_bid, filter, n)
+      when is_integer(sid) and is_integer(n) and filter in [:pending, :all, :ingested] do
+    # chunk over table and drop
+    ms =
+      Ex2ms.fun do
+        {_event_id, _status, _event} = obj when ^filter == :all -> obj
+        {_event_id, status, _event} = obj when status == ^filter -> obj
+      end
+
+    with tid when tid != nil <- get_tid(sid_bid) do
+      truncate_traverse(tid, :ets.select(tid, ms, 100), 0, n)
+    else
+      nil -> {:error, :not_initialized}
+      :"$end_of_table" -> {:ok, []}
+    end
+  end
+
+  defp truncate_traverse(_tid, :"$end_of_table", _acc, _limit), do: :ok
+
+  defp truncate_traverse(tid, {taken, cont}, acc, limit) when acc < limit do
+    # keep
+    diff = limit - acc
+    rem = Enum.count(taken) - diff
+
+    to_add =
+      if rem > 0 do
+        # satisfy all, drop up to required
+        for {key, _status, _event} <- Enum.take(taken, diff) do
+          :ets.delete(tid, key)
+        end
+
+        diff
+      else
+        # cannot satisfy all, keep all and continue
+        Enum.count(taken)
+      end
+
+    truncate_traverse(tid, :ets.select(cont), acc + to_add, limit)
+  end
+
+  defp truncate_traverse(tid, {taken, cont}, acc, limit) when acc >= limit do
+    # delete the rest
+    for {key, _status, _event} <- taken do
+      :ets.delete(tid, key)
+    end
+
+    truncate_traverse(tid, :ets.select(cont), acc, limit)
+  end
+
+  def delete_all_mappings do
+    :ets.delete_all_objects(@ets_table_mapper)
+  end
+
+  def delete_stale_mappings do
+    ms =
+      Ex2ms.fun do
+        obj -> obj
+      end
+
+    :ets.select(@ets_table_mapper, ms, 100)
```

Comment:
maybe with match is better 
```suggestion
    ets.match(@ets_table_mapper, :'$1', 100)
```